add cpu limit to task and container

tasnimzotder opened the pull request:
## Summary

  Added CPU resource management for ECS services with automatic limit calculation to improve service reliability and resource allocation.

  ### New Features:
  - `cpu_reservation` field for EC2 launch type (128-4096 CPU units, optional)
  - Automatic CPU limit calculation (125% of reservation for burst capacity)
  - Enhanced service name validation (now supports numbers: `Service1`, `Kafdrop2`)

  ### Benefits:
  - 25% CPU headroom prevents service throttling during traffic spikes
  - Better ECS task scheduling with predictable resource requirements
  - Improved cluster capacity planning and cost optimization

  ### Backward Compatibility:
  - ✅ Zero breaking changes - `cpu_reservation` is optional
  - ✅ All existing services continue working unchanged

  ### Technical Details:
  - When `cpu_reservation` is set (e.g., 512), task gets 640 CPU limit (125%); 512x125% = 640.
  - Container CPU remains at 0 (unchanged from existing behavior)
  - When cpu_reservation is None, task CPU remains None (existing behavior)
  - JSON schema validation prevents invalid configurations
  - Only applies to EC2 launch type (Fargate uses existing fargate.cpu)

tasnimzotder committed:
add cpu limit to task and container
Added to the service config:
- CPU Reservation
- CPU Limit

Both of the values are not required. CPU Reservation is recommended.

Co-authored-by: Rokibul Islam <hello@rokibul.dev>

tasnimzotder committed:
remove redundant test_aws_credentials fixture

tasnimzotder committed:
remove cpu_limit, cap cpu_reservation at 4096

Removed cpu_limit parameter to avoid confusion.
Set a max value of 4096 units (4 vCPU) for cpu_reservation
to prevent users from assigning higher values

tasnimzotder committed:
remove redundant test cases

tasnimzotder committed:
fix task cpu limit string

In task level, the CPU limit needs to be whole number
without any decimal value in string format

rokibulislaam commented on cloudlift/deployment/service_template_generator.py:
```diff
@@ -233,8 +233,14 @@ def _add_service(self, service_name, config):
             "Image": self.ecr_image_uri + ':' + self.current_version,
             "Essential": 'true',
-            "Cpu": 0
+            "Cpu": 0 # Default to 0
```
Why is the default value set to `0`? I'm curious about the impact on existing services where the CPU reservation isn't explicitly specified. There could be several reasons for this, for example, users might not have visibility into CPU utilization metrics or might not know how much CPU the container or the overall task actually needs.

According to the [AWS documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinition.html#cfn-ecs-taskdefinition-containerdefinition-cpu), this configuration maps to Docker’s `--cpu-shares` option. The same document notes:

> Agent versions greater than or equal to 1.2.0: Null, zero, and CPU values of 1 are passed to Docker as 2.

However, [Docker’s documentation](https://docs.docker.com/engine/containers/resource_constraints/#cpu) states that the default value for `--cpu-shares` is 1024. CPU shares are relative. Given that, wouldn’t setting the value to `2` (as the ECS agent does in this case) result in disproportionately low CPU allocation during contention, potentially leading to unexpected performance issues?

Wouldn't it be better to leave the value unset and let Docker apply its default? What would happen in that case?

This needs a discussion.

rokibulislaam replied:
After a discussion, we decided not to mention any value to "Cpu" of container definition.
Going to retain the default 0 so that existing functionalities don't break.

rokibulislaam commented on requirements.dev.txt:
```diff
@@ -1,2 +1,3 @@
-pytest==4.0.0
\ No newline at end of file
+moto==5.1.4
+pytest==4.0.0
```
let's upgrade pytest.

rokibulislaam commented on test/env.sample:
Is this file required?

tasnimzotder committed:
remove container level cpu reservation

Details in https://github.com/GetSimpl/cloudlift/pull/163#discussion_r2084364353

tasnimzotder committed:
add use_container_cpu_reservation flag

shoan commented on cloudlift/utils.py:
```diff
@@ -10,3 +14,36 @@ def generate_pascalcase_name(name: str, max_length: int = 32) -> str:
     return pascalcase_name
+
+
+def flatten_dict(d, parent_key="", sep="."):
```
If this function is only being used in the tests, it should be part of the test suite's utils.

rokibulislaam replied:
Yes, for now this is only being used in test cases. This can be moved to test suite's utils.

shoan commented on cloudlift/config/service_configuration.py:
```diff
@@ -240,9 +240,21 @@ def _validate_changes(self, configuration):
+                    "minimum": 128,
+                    "maximum": 4096
+                },
+                "use_container_cpu_reservation": {
```
Cloudlift is an opinionated abstraction over ecs/docker. As a user, I don't know how these two differ. I can relate to soft limit and hard limit. I don't intuitively understand the difference between the two cpu parameters.

rokibulislaam replied:
The parameter `use_container_cpu_reservation` has been removed.

shoan commented on cloudlift/deployment/service_template_generator.py:
```diff
@@ -235,6 +235,11 @@ def _add_service(self, service_name, config):
             "Cpu": 0
         }
+
+        if config.get("cpu_reservation", None) and config.get("use_use_container_cpu_reservation", False):
```
Is this intentional `use_use_...`?

rokibulislaam commented:
`deploy_service` is removing the "cpu" field from task definition. @tasnimzotder we need to fix that.

tasnimzotder committed:
update deploy_service for cpu_reservtion

rokibulislaam commented:
@tasnimzotder please bump version to v2.2.0.

tasnimzotder committed:
bump version to 2.2.0

tasnimzotder committed:
allow numbers in service names while maintaining alphanumeric-only validation

- Update service name regex pattern from ^[a-zA-Z]+$ to ^[a-zA-Z][a-zA-Z0-9]*$
- Service names must now start with a letter and can contain letters and numbers
- Update comprehensive test suite with alphanumeric service name patterns

tasnimzotder committed:
fix test cases for service_configuration

tasnimzotder committed:
remove duplicate test cases

shoan commented on cloudlift/deployment/ecs.py:
```diff
@@ -424,22 +428,26 @@ def get_task_definition(self, task_definition):
             execution_role_arn=task_definition.execution_role_arn if task_definition.execution_role_arn else boto3.resource('iam').Role('ecsTaskExecutionRole').arn,
             network_mode=task_definition.network_mode or u'bridge',
-            **fargate_td
+            **base_td
```
Retain specific overrides in `fargate_td` . Move this configuration into `base_td` and apply overrides to it as configured in the service.

shoan commented on cloudlift/deployment/service_template_generator.py:
```diff
@@ -332,6 +333,25 @@ def _add_service(self, service_name, config):
+                cpu_limit = cpu_reservation * multiplier_factor
+
+            # Do not set a default value unless cpu_reservation is provided
+            if cpu_limit:
```
Be explicit in this check.

tasnimzotder committed:
refactor: enhance task definition update logic and add utility functions for testing

tasnimzotder committed:
refactor: introduce CPU limit multiplier constant for better readability and maintainability

rokibulislaam commented on cloudlift/deployment/ecs.py:
```diff
@@ -424,23 +428,30 @@ def get_task_definition(self, task_definition):
-            network_mode=task_definition.network_mode or u'bridge',
-            **fargate_td
-        )
+            base_td.update(fargate_td)
```
Will python dict.update() handle deep merging? We also need to handle deep merging.

For example, in case there was a docker mount point added, and now we add an EFS mount point with cloudlift, in this case, dict.update will just remove the docker mount point. Example:


```
container_def = {
    "volumes": [
        {
            "name": "test-efs-vol",
            "efsVolumeConfiguration": {
                "fileSystemId": "fs-abcd1234",
                "rootDirectory": "/",
            },
        }
    ],
}

container_def.update(
    {
        "volumes": [
            {
                "name": "test-docker-vol",
                "dockerVolumeConfiguration": {"scope": "task", "driver": "local"},
            }
        ]
    }
)

print(container_def)

# Print output:
# {
#     "volumes": [
#         {
#             "name": "test-docker-vol",
#             "dockerVolumeConfiguration": {"scope": "task", "driver": "local"},
#         }
#     ]
# }

```

tasnimzotder replied:
This PR introduces task-level CPU field support but does not preserve all existing task definition fields during `deploy_service` operations. Deep merge functionality is intentionally not included in this release. We plan to address comprehensive task definition field preservation in a future release, which will include respecting manual changes made through the AWS Console.

rokibulislaam reviewed (commented):
Please have a look @tasnimzotder
